doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+the priority of inlining the call. It makes it more likely to inline the call because there might be other
+optimizations that can happen downstream because of passing in this constant `String`. But inlining will not do
+that optimization, that optimization will be done separately by other optimization paths when they run.
+
+Generally speaking, other optimization do not differentiate if the trees are inlined calls or not.
+
+Occasionally, some optimizations might be interested in where the code comes from. For example, if we know a method
+is from `String` class, we know all the access of the method will be within the array bound. It is impossible to
+raise an exception there. The bound check can be skipped for the method that comes from String class. Such thing
+is normally restricted to sets of Java class library methods. Even if the `String` method is inlined into some
+other arbitrary method `foo()`, that property should not change. The bound check should not be left in place
+simply because `foo()` is being compiled. If it comes from the `String` code, regardless of whether it is inlined
+or not, we should still be able to omit that bound check. This is an example of an awareness that the code comes
+from an inlined method and where we differentiate the inlined code.
+
+There are other enhancements that have been done over the years, such as method handle inlining. There is code
+that analyzes the callee’s bytecode and looks at whether there are opportunities to do some optimization. More
+and more code has been added in Inliner to be incrementally aware of the opportunities when going into the callees.
+
+# 4. Overview Of Guards
+
+## Hot Code Replacement (HCR) Guard
+
+HCR is on by default. Any Java method can be redefined by the user while the program is running. As a result,
+we must be able to reflect the output when any of the redefinitions happen. Even if there is a direct call such
+as `invoke_static` or `invoke_special`, a guard must be added around it to preserve the functional correctness
+in case that the method is redefined. The HCR guard works as a nopable guard (discussed in the vnext section).
+The checking is not done explicitly in the generated code, which is good since there will be no overhead in doing
+the extra check. If HCR is not enabled, a direct call is inlined without any guard.
+
+HCR guards can be treated as a second guard. If there is no other guard, it introduces a new guard. If there is
+a guard, it adds another layer as a second guard. For example, it is like two conditions, a virtual guard and
+a HCR guard both exist in place to indicate if the method is overridden, or if the method is redefined.
+
+## Nopable Guard
+
+Nopable guard is like a conditional test conceptually except there is no explicit testing in the generated code.
+There is an `“if”` in IL, but it will always fall into the inlined code during the compilation. If the user
```

Comment:
"it will always fall into the inlined code during the compilation" -> "it is almost always expected to fall into the inlined code just after the compilation is done"